neuraxle/base.py

```diff
@@ -126,6 +98,35 @@ def __init__(
 
         self.pending_mutate: ('BaseStep', str, str) = (None, None, None)
         self.is_initialized = False
+        self.parent_step = None
+
+    def hash(self, current_ids, hyperparameters, data_inputs: Any = None):
+        if current_ids is None:
+            current_ids = [str(i) for i in range(len(data_inputs))]
+
+        if len(hyperparameters) == 0:
+            return current_ids
+
+        current_hyperparameters_hash = self._hash_hyperparameters(hyperparameters)
+
+        new_current_ids = []
+        for current_id in current_ids:
+            m = hashlib.md5()
+            m.update(str.encode(current_id))
+            m.update(str.encode(current_hyperparameters_hash))
+            new_current_ids.append(m.hexdigest())
+
+        return new_current_ids
+
+    def _hash_hyperparameters(self, hyperparams: HyperparameterSamples):
+        hyperperams_dict = hyperparams.to_flat_as_dict_primitive()
+        return hashlib.md5(str.encode(str(hyperperams_dict))).hexdigest()
+
+    def set_parent(self, parent_step: 'BaseStep'):
```

Comment:
kind of like the composite design pattern : https://sourcemaking.com/design_patterns/composite
but with a way for accessing a component's parent in the recursive structure...

Reply:
maybe we should set a callback function instead of the whole parent object reference

Reply:
`Pipeline`, `TruncableSteps` and `MetaStepMixin` are already composites. 

For access to global properties as you suggest, your reference says this: 
> Composite could use Chain of Responsibility to let components access global properties through their parent. It could also use Decorator to override these properties on parts of the composition.

This is interesting, supporting having a link to the parent: 
https://stackoverflow.com/a/38740454/2476920

I realize that we perhaps need the `apply` method more than we think we need. Issue #66 is for that apply.

Reply:
Note that in 90% of the time (or more) it seems that the tree is traversed only from root to leaf childs, and it is rarily going back up to the parents.

Reply:
90% people don't need to go up the tree, I think that is why.

Reply:
Composite could use Chain of Responsibility to let components access global properties through their parent : https://sourcemaking.com/design_patterns/chain_of_responsibility

Reply:
I suggest creating a context object where the entire chain will lie. See #69.

Reply:
Because I added an issue, you can resolve this.